--- a/modules.d/99base/module-setup.sh 
+++ b/modules.d/99base/module-setup.sh
@@ -74,6 +73,13 @@
         inst_multiple switch_root || dfatal "Failed to install switch_root"
         inst_script "$moddir/init.sh" "/init"
         inst_hook cmdline 10 "$moddir/parse-root-opts.sh"
+
+        {
+            echo "NAME=dracut"
+            echo "ID=dracut"
+            echo "VERSION_ID=\"$DRACUT_VERSION\""
+            echo 'ANSI_COLOR="0;34"'
+        } > "${initdir}"/usr/lib/initrd-release
     fi
 
     ln -fs /proc/self/mounts "$initdir/etc/mtab"
@@ -81,13 +87,6 @@
         echo ro >> "${initdir}/etc/cmdline.d/base.conf"
     fi
 
-    {
-        echo "NAME=dracut"
-        echo "ID=dracut"
-        echo "VERSION_ID=\"$DRACUT_VERSION\""
-        echo 'ANSI_COLOR="0;34"'
-    } > "${initdir}"/usr/lib/initrd-release
-
     echo "dracut-$DRACUT_VERSION" > "$initdir/lib/dracut/dracut-$DRACUT_VERSION"
 
     ## save host_devs which we need bring up
