diff --git a/modules.d/10systemd/module-setup.sh b/modules.d/10systemd/module-setup.sh
index a937cd1a..16a007a8 100755
--- a/modules.d/10systemd/module-setup.sh
+++ b/modules.d/10systemd/module-setup.sh
@@ -155,5 +155,6 @@ EOF
     # Install library file(s)
     inst_libdir_file \
         {"tls/$DRACUT_ARCH/",tls/,"$DRACUT_ARCH/",}"libbpf.so*" \
-        {"tls/$DRACUT_ARCH/",tls/,"$DRACUT_ARCH/",}"libnss_*"
+        {"tls/$DRACUT_ARCH/",tls/,"$DRACUT_ARCH/",}"libnss_*" \
+        {"tls/$DRACUT_ARCH/",tls/,"$DRACUT_ARCH/",}"libseccomp_*"
 }
